### PR TITLE
Users

### PR DESCRIPTION
**Pull request by smagolexandr**
* add user edit
* add user search

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -44,7 +44,7 @@
 /composer.phar
 
 # Backup entities generated with doctrine:generate:entities command
-*/Entity/*~
+*~
 
 # PhpStorm and Mac fs
 .idea
```
Comment (ahonymous) [inline .gitignore]: add this for global gitignore

File: app/Resources/views/default/index.html.twig [hunk 2]
```diff
@@ -13,8 +13,13 @@
...
                 <li><a href="#">Request forms</a></li>
                 <li><a href="#">Surveys</a></li>
-                <li><a href="{{ path('user_list') }}">Users</a></li>
-                <li><a href="#">Sign in</a></li>
+                <li><a href="{{ path('users_list') }}">Users</a></li>
+                {% if app.user %}
+                    <li><a href="#">Hi, {{ app.user.userName }}</a></li>
+                    <li><a href="{{ path('logout') }}">Sign out</a></li>
+                {% else %}
+
```
Comment (spolischook) [inline app/Resources/views/default/index.html.twig]: Maybe link to login?

File: src/AppBundle/Controller/UserController.php [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+     * @Template("@App/users.html.twig")
+     *
+     * @param Request $request
+     * @return array
+     */
+    public function usersListAction(Request $request)
+    {
+        $em = $this->getDoctrine()->getManager();
+        $users = $em->getRepository(User::class)->getUsersByParams($request->query);
+        if (!$users) {
```
Comment (ahonymous) [inline src/AppBundle/Controller/UserController.php]: if you have not found any user you need return the empty array

Comment (ahonymous) [inline src/AppBundle/Controller/UserController.php]: better your repository find method has return the query builder for paginator

File: src/AppBundle/Controller/UserController.php [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        $paginator = $this->get('knp_paginator');
+        $pagination = $paginator->paginate($users, $request->query->getInt('page', 1), 10);
+        return [
+            "users" => $pagination
+        ];
+    }
+
+    /**
+     * @Route("/user/activate/{id}", name="activate_user")
+     * @Template("@App/add.html.twig")
```
Comment (ahonymous) [inline src/AppBundle/Controller/UserController.php]: what is the method in this route?

File: src/AppBundle/Controller/UserController.php [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+     * @return RedirectResponse
+     */
+    public function userActivateAction(User $user)
+    {
+        $em = $this->getDoctrine()->getManager();
+        $user->setStatus($user->isEnabled() ? false : true);
+        $em->persist($user);
+        $em->flush();
+
+        return new RedirectResponse($this->generateUrl("users_list"));
```
Comment (ahonymous) [inline src/AppBundle/Controller/UserController.php]: use the "$this->redirect(...)" or "$this->redirectToRoute(...)"

File: src/AppBundle/Controller/UserController.php [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+     * @Route("/user/activate/{id}", name="activate_user")
+     * @Template("@App/add.html.twig")
+     *
+     * @param  User $user
+     * @return RedirectResponse
+     */
+    public function userActivateAction(User $user)
+    {
+        $em = $this->getDoctrine()->getManager();
+        $user->setStatus($user->isEnabled() ? false : true);
```
Comment (ahonymous) [inline src/AppBundle/Controller/UserController.php]: I think that this flow better implement with form

File: src/AppBundle/Controller/UserController.php [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+     * @param Request $request
+     * @return array|RedirectResponse
+     */
+    public function userAddAction(Request $request)
+    {
+        $em = $this->getDoctrine()->getManager();
+        $user = new User();
+        $form = $this->createForm('AppBundle\Form\UserType', $user, [
+            'action' => $this->generateUrl('add_user'),
+            'method' => 'POST',
```
Comment (ahonymous) [inline src/AppBundle/Controller/UserController.php]: by default the form has a method option as POST

File: src/AppBundle/Controller/UserController.php [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+            'validation_groups' => array('registration'),
+        ])
+            ->add('Save', SubmitType::class, array(
```
Comment (ahonymous) [inline src/AppBundle/Controller/UserController.php]: good practice was add buttons to the form in the template

File: src/AppBundle/Controller/UserController.php [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+                'attr' => ['class' => 'btn pull-right btn-warning']
+            ));
+        $form->handleRequest($request);
+        if ($form->isValid()) {
+            $em->persist($user);
+            $em->flush();
+
+            return new RedirectResponse($this->generateUrl("users_list"));
```
Comment (ahonymous) [inline src/AppBundle/Controller/UserController.php]: use the "$this->redirect(...)" or "$this->redirectToRoute(...)"

File: src/AppBundle/DataFixtures/ORM/serve_seatle_fixtures.yml [hunk 10]
```diff
@@ -1,23 +1,26 @@
...
+      username: admin
+      email: admin@serve-seattle.com
+      plainPassword: admin
+
 AppBundle\Entity\User:
     intern1:
         firstName: <firstName()>
+        lastName: <lastName()>
         email: <email()>
         plainPassword: intern1
```
Comment (ahonymous) [inline src/AppBundle/DataFixtures/ORM/serve_seatle_fixtures.yml]: how you fixtures coded the password?

File: app/Resources/views/default/index.html.twig (hunk 2 above)
Comment (smagolexandr) [inline app/Resources/views/default/index.html.twig]: u can't see menu until you log in

File: src/AppBundle/DataFixtures/ORM/serve_seatle_fixtures.yml (hunk 10 above)
Comment (smagolexandr) [inline src/AppBundle/DataFixtures/ORM/serve_seatle_fixtures.yml]: we have passwordlistener for it

File: src/AppBundle/Entity/User.php [hunk 11]
```diff
@@ -44,6 +46,7 @@ class User implements AdvancedUserInterface, \Serializable
 
     /**
      * @var string
+     * @Assert\NotBlank()
      * @Assert\Regex(
      *     pattern="/\d/",
```
Comment (ahonymous) [inline src/AppBundle/Entity/User.php]: you have a string with constraint for digits?

File: src/AppBundle/Entity/User.php [hunk 12]
```diff
@@ -79,17 +83,17 @@ class User implements AdvancedUserInterface, \Serializable
...
      */
     private $password;
 
     /**
      * @var string
+     * @Assert\Type("string")
+     * @Assert\Length(
+     *      max = 255
+     * )
+     * @Assert\NotBlank(groups={"registration"})
```
Comment (ahonymous) [inline src/AppBundle/Entity/User.php]: Will be a group for forgot password?

File: src/AppBundle/Form/UserType.php [hunk 13]
```diff
@@ -78,7 +71,8 @@ public function buildForm(FormBuilderInterface $builder, array $options)
     public function configureOptions(OptionsResolver $resolver)
     {
         $resolver->setDefaults(array(
-            'data_class' => 'AppBundle\Entity\User'
+            'data_class' => 'AppBundle\Entity\User',
```
Comment (ahonymous) [inline src/AppBundle/Form/UserType.php]: replace 'AppBundle\Entity\User' on User::class

File: src/AppBundle/Controller/UserController.php (hunk 3 above)
Comment (smagolexandr) [inline src/AppBundle/Controller/UserController.php]: Do i need to move 
`$paginator = $this->get('knp_paginator');
        $pagination = $paginator->paginate($users, $request->query->getInt('page', 1), 10);`
to repo?

File: src/AppBundle/Entity/User.php [hunk 14]
```diff
@@ -305,7 +309,7 @@ public function setPlainPassword($plainPassword)
      *
      * @return $this
      */
-    public function setIsEnabled($status)
+    public function setStatus($status)
     {
         $this->isActive = $status;
```
Comment (ahonymous) [inline src/AppBundle/Entity/User.php]: Why property name for enabling user has name as verb?

File: src/AppBundle/Controller/UserController.php (hunk 4 above)
Comment (smagolexandr) [inline src/AppBundle/Controller/UserController.php]: i think "put", isn't it?

File: src/AppBundle/Repository/UserRepository.php [hunk 15]
```diff
@@ -20,4 +21,30 @@ public function loadUserByUsername($username)
...
+
+    /**
+     * @param ParameterBag $params
+     * @return array
+     */
+    public function getUsersByParams(ParameterBag $params):array
+    {
+        $em = $this->getEntityManager();
+
+        $postsQuery = $em->createQueryBuilder()
```
Comment (ahonymous) [inline src/AppBundle/Repository/UserRepository.php]: replace $em->createQueryBuilder() on  $this->createQueryBuilder()

File: src/AppBundle/Entity/User.php (hunk 14 above)
Comment (smagolexandr) [inline src/AppBundle/Entity/User.php]: it is AdvancedUserInterface :c

File: src/AppBundle/Controller/UserController.php (hunk 3 above)
Comment (ahonymous) [inline src/AppBundle/Controller/UserController.php]: no I need see 
 $paginator->paginate($em->getRepository(User::class)->getUsersByParams($request->query), $request->query->getInt('page', 1), 10)

and you "getUsersByParams" should named as "selectUsersByParams"

File: src/AppBundle/Entity/User.php (hunk 12 above)
Comment (smagolexandr) [inline src/AppBundle/Entity/User.php]: we haven't realized it yet.

File: app/Resources/views/default/index.html.twig (hunk 2 above)
Comment (spolischook) [inline app/Resources/views/default/index.html.twig]: So why you need this ```if``` statement?

File: src/AppBundle/Controller/UserController.php (hunk 6 above)
Comment (spolischook) [inline src/AppBundle/Controller/UserController.php]: ```$user->isEnabled()``` must return bool, add typehinting and replace this to:
```php
$user->setStatus($user->isEnabled());
```

Comment (spolischook) [inline src/AppBundle/Controller/UserController.php]: however the code:
```php
$user->setStatus($user->isEnabled());
```
looks weird

File: app/Resources/views/default/index.html.twig (hunk 2 above)
Comment (smagolexandr) [inline app/Resources/views/default/index.html.twig]: fixed

File: src/AppBundle/Controller/UserController.php (hunk 6 above)
Comment (smagolexandr) [inline src/AppBundle/Controller/UserController.php]: it return true if user is enabled. ps It is property from AdvancedUserInterface

File: app/Resources/views/default/index.html.twig (hunk 2 above)
Comment (spolischook) [inline app/Resources/views/default/index.html.twig]: What are you fixing? I still see the ```{% if app.user %}```

File: src/AppBundle/Controller/UserController.php [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+     * @Route("/user/activate", name="activate_user")
+     * @Template("@App/add.html.twig")
+     *
+     * @Method("PUT")
+     * @param  Request $request
+     * @return RedirectResponse
+     */
+    public function userActivateAction(Request $request)
+    {
+        $em = $this->getDoctrine()->getManager();
```
Comment (ahonymous) [inline src/AppBundle/Controller/UserController.php]: Where is this handle of activate form?

File: src/AppBundle/Entity/User.php (hunk 11 above)
Comment (lilia86) [inline src/AppBundle/Entity/User.php]: see below 'match=false'

File: src/AppBundle/Controller/UserController.php (hunk 16 above)
Comment (smagolexandr) [inline src/AppBundle/Controller/UserController.php]: it is handler

File: src/AppBundle/Controller/UserController.php [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+        $form = $this->createForm(ActivationType::class, $user, [
+            'method' => "PUT",
+            'action' => $this->generateUrl('activate_user', array('id' => $user->getId())),
+            'validation_groups' => array('edit'),
+        ]);
+        $form->handleRequest($request);
+        if ($form->isValid()) {
+            $em->persist($user);
+            $em->flush();
+            return $this->redirect($this->generateUrl("users_list"));
```
Comment (ahonymous) [inline src/AppBundle/Controller/UserController.php]: Where is your code style?

File: src/AppBundle/Controller/UserController.php [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+        }
+        return $this->redirect($this->generateUrl("users_list"));
```
Comment (ahonymous) [inline src/AppBundle/Controller/UserController.php]: Where is your code style?

File: src/AppBundle/Form/User/EditType.php [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+use Symfony\Component\Form\Extension\Core\Type\RepeatedType;
+use Symfony\Component\Form\Extension\Core\Type\PasswordType;
+use AppBundle\Entity\User;
+
+/**
+ * Class UserType
+ * @package AppBundle\Form
+
+ */
+class EditType extends AbstractType
```
Comment (ahonymous) [inline src/AppBundle/Form/User/EditType.php]: Maybe your mean this class name as UserType?

File: src/AppBundle/Repository/UserRepository.php [hunk 20]
```diff
@@ -20,4 +21,28 @@ public function loadUserByUsername($username)
...
+    /**
+     * @param ParameterBag $params
+     * @return array
+     */
+    public function selectUsersByParams(ParameterBag $params):array
+    {
+        $postsQuery = $this->createQueryBuilder('Users')
+            ->select('u')
+            ->from('AppBundle:User', 'u')
+            ->orderBy('u.createdAt', preg_match('/asc/i', $params->get('order')) ? 'ASC' : 'DESC');
```
Comment (ahonymous) [inline src/AppBundle/Repository/UserRepository.php]: Validation and transformation of request should be in controller!

File: src/AppBundle/Resources/views/users.html.twig [hunk 21]
```diff
@@ -7,22 +7,37 @@
 {% block content %}
     <h1>
         Users
-        <small>
+    </h1>
+    <div>
+        <form method="get" class="filters">
```
Comment (ahonymous) [inline src/AppBundle/Resources/views/users.html.twig]: create form on the filter and bind the query string for DTO filter

File: src/AppBundle/Resources/views/users.html.twig [hunk 22]
```diff
@@ -7,22 +7,37 @@
...
+            <i class="fa fa-filter" aria-hidden="true"></i>
+            <select name="order">
+                <option {% if app.request.get('order') == 'desc' or not app.request.get('order') %}selected{% endif %} value="desc">
```
Comment (ahonymous) [inline src/AppBundle/Resources/views/users.html.twig]: ordering of user can do the paginator

File: src/AppBundle/Form/User/EditType.php (hunk 19 above)
Comment (smagolexandr) [inline src/AppBundle/Form/User/EditType.php]: check file destination

**Pull request merged**